// @antv/l7-draw/es/control/index.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t.return && (u = t.return(), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _get() { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get.bind(); } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(arguments.length < 3 ? target : receiver); } return desc.value; }; } return _get.apply(this, arguments); }
function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
import { Control, DOM } from '@antv/l7';
import { DrawBtnActiveClassName, DrawBtnClassName, DrawControlClassName, DrawIconMap, DrawInstanceMap, DrawTypeAttrName, DrawTypeList } from './constant';
import { getParentByClassName } from '../typings';
import './iconfont.js';
import "./index.css";
import { ControlEvent, DrawEvent } from '../constant';
import { debounce, fromPairs, omit, toPairs } from 'lodash';
export var DrawControl = /*#__PURE__*/function (_Control) {
  _inherits(DrawControl, _Control);
  var _super = _createSuper(DrawControl);
  function DrawControl(scene) {
    var _this;
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    _classCallCheck(this, DrawControl);
    _this = _super.call(this, options);
    _this.controlOption = void 0;
    _this.sceneInstance = void 0;
    _this.drawMap = {};
    _this.btnMap = {};
    _this.activeType = null;
    _this.emitDataChange = debounce(function () {
      _this.emit(ControlEvent.DataChange, _this.getDrawData());
    }, 16);
    _this.emitDrawSelect = function (drawType, feature) {
      _this.emit(ControlEvent.DrawSelect, drawType, feature);
    };
    _this.sceneInstance = scene;
    _this.controlOption = _objectSpread(_objectSpread({}, _this.getDefault()), options || {});
    _this.onBtnClick = _this.onBtnClick.bind(_assertThisInitialized(_this));
    return _this;
  }
  _createClass(DrawControl, [{
    key: "getActiveType",
    value: function getActiveType() {
      return this.activeType;
    }
    // @ts-ignore
  }, {
    key: "getDefault",
    value: function getDefault() {
      return _objectSpread(_objectSpread({}, _get(_getPrototypeOf(DrawControl.prototype), "getDefault", this).call(this)), {}, {
        className: '',
        buttonClassName: '',
        activeButtonClassName: '',
        style: '',
        position: 'topleft',
        commonDrawOptions: {},
        drawConfig: {
          point: true,
          line: true,
          polygon: true,
          rect: true,
          circle: true,
          clear: true
        }
      });
    }
  }, {
    key: "addTo",
    value: function addTo(container) {
      _get(_getPrototypeOf(DrawControl.prototype), "addTo", this).call(this, container);
      this.init();
      return this;
    }
  }, {
    key: "onAdd",
    value: function onAdd() {
      var _this$controlOption = this.controlOption,
        className = _this$controlOption.className,
        style = _this$controlOption.style;
      var container = DOM.create('div', "".concat(DrawControlClassName, " ").concat(className));
      if (style) {
        container.setAttribute('style', style);
      }
      return container;
    }
  }, {
    key: "onRemove",
    value: function onRemove() {
      Object.values(this.drawMap).forEach(function (draw) {
        draw.clear(true);
      });
    }
  }, {
    key: "init",
    value: function init() {
      var _this2 = this,
        _this$container;
      var btnList = [];
      Object.entries(this.controlOption.drawConfig).filter(function (_ref) {
        var _ref2 = _slicedToArray(_ref, 2),
          _ = _ref2[0],
          options = _ref2[1];
        return options;
      }).sort(function (item1, item2) {
        var _item = _slicedToArray(item1, 2),
          options1 = _item[1];
        var _item2 = _slicedToArray(item2, 2),
          options2 = _item2[1];
        var MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER;
        var order1 = typeof options1 === 'boolean' ? MAX_SAFE_INTEGER : options1.order;
        var order2 = typeof options2 === 'boolean' ? MAX_SAFE_INTEGER : options2.order;
        return order1 - order2;
      }).map(function (_ref3) {
        var _ref4 = _slicedToArray(_ref3, 2),
          drawType = _ref4[0],
          options = _ref4[1];
        return [drawType, _typeof(options) === 'object' ? omit(options, 'order') : options];
      }).forEach(function (_ref5) {
        var _ref6 = _slicedToArray(_ref5, 2),
          btnType = _ref6[0],
          options = _ref6[1];
        if (_typeof(options) === 'object' && options.button) {
          var btn = options.button;
          if (!btn.classList.contains('l7-draw-control__btn')) {
            btn.classList.add('l7-draw-control__btn');
          }
          _this2.btnMap[btnType] = btn;
          btnList.push(btn);
        } else if (options && DrawTypeList.includes(btnType)) {
          var newBtn = _this2.initBtn(btnType);
          newBtn.addEventListener('click', _this2.onBtnClick);
          btnList.push(newBtn);
          _this2.btnMap[btnType] = newBtn;
          // @ts-ignore;
          var Draw = DrawInstanceMap[btnType];
          if (Draw && _this2.sceneInstance) {
            var draw = new Draw(_this2.sceneInstance, _objectSpread(_objectSpread({}, _this2.controlOption.commonDrawOptions), typeof options === 'boolean' ? {} : options));
            draw.on(DrawEvent.Change, _this2.emitDataChange);
            draw.on(DrawEvent.Select, function (feature) {
              _this2.emitDrawSelect(btnType, feature);
            });
            _this2.drawMap[btnType] = draw;
          }
        }
      });
      (_this$container = this.container).append.apply(_this$container, btnList);
      if (this.controlOption.defaultActiveType) {
        this.onDrawClick(this.controlOption.defaultActiveType);
      }
    }
  }, {
    key: "getDrawData",
    value: function getDrawData() {
      return fromPairs(toPairs(this.drawMap).map(function (_ref7) {
        var _ref8 = _slicedToArray(_ref7, 2),
          drawType = _ref8[0],
          draw = _ref8[1];
        return [drawType, draw.getData()];
      }));
    }
  }, {
    key: "getTypeDraw",
    value: function getTypeDraw(type) {
      var _this$drawMap$type;
      return (_this$drawMap$type = this.drawMap[type]) !== null && _this$drawMap$type !== void 0 ? _this$drawMap$type : null;
    }
  }, {
    key: "onBtnClick",
    value:
    /**
     * 按钮的点击事件
     * @param e
     */
    function onBtnClick(e) {
      var btn = getParentByClassName(e.target, DrawBtnClassName);
      if (!btn) {
        return;
      }
      var btnType = btn.getAttribute(DrawTypeAttrName);
      if (btnType === 'clear') {
        this.clearDrawData();
      } else {
        this.onDrawClick(btnType);
      }
    }
    /**
     * 设置当前激活的绘制类型
     * @param type
     */
  }, {
    key: "setActiveType",
    value: function setActiveType(type) {
      var newActiveType = null;
      var oldActiveType = this.activeType;
      if (oldActiveType) {
        var _this$drawMap$oldActi;
        var targetBtn = this.btnMap[oldActiveType];
        if (targetBtn) {
          DOM.removeClass(targetBtn, DrawBtnActiveClassName);
          var activeButtonClassName = this.controlOption.activeButtonClassName;
          if (activeButtonClassName) {
            DOM.removeClass(targetBtn, activeButtonClassName);
          }
        }
        (_this$drawMap$oldActi = this.drawMap[oldActiveType]) === null || _this$drawMap$oldActi === void 0 ? void 0 : _this$drawMap$oldActi.disable();
        newActiveType = null;
      }
      if (oldActiveType !== type && type) {
        var _this$drawMap$type2;
        (_this$drawMap$type2 = this.drawMap[type]) === null || _this$drawMap$type2 === void 0 ? void 0 : _this$drawMap$type2.enable();
        var _targetBtn = this.btnMap[type];
        _targetBtn && DOM.addClass(_targetBtn, "".concat(DrawBtnActiveClassName, " ").concat(this.controlOption.activeButtonClassName));
        newActiveType = type;
      }
      this.activeType = newActiveType;
      this.emit(ControlEvent.DrawChange, newActiveType);
    }
    /**
     * 绘制按钮点击的回调
     * @param type
     */
  }, {
    key: "onDrawClick",
    value: function onDrawClick(type) {
      this.setActiveType(type);
    }
    /**
     * 清空绘制数据
     */
  }, {
    key: "clearDrawData",
    value: function clearDrawData() {
      Object.values(this.drawMap).forEach(function (draw) {
        draw === null || draw === void 0 ? void 0 : draw.clear();
      });
      this.emit(ControlEvent.DrawClear);
    }
    /**
     * 创建按钮
     * @param type
     * @returns
     */
  }, {
    key: "initBtn",
    value: function initBtn(type) {
      var btn = DOM.create('button', "".concat(DrawBtnClassName, " ").concat(this.controlOption.buttonClassName));
      btn.innerHTML = "<svg class=\"l7-draw-icon\" aria-hidden=\"true\">\n      <use xlink:href=\"".concat(DrawIconMap[type], "\"></use>\n    </svg>");
      btn.setAttribute(DrawTypeAttrName, type);
      return btn;
    }
  }]);
  return DrawControl;
}(Control);
