// @antv/l7-draw/es/render/line-render.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
import { LineLayer } from '@antv/l7';
import { featureCollection } from '@turf/turf';
import { debounce } from 'lodash';
import { LayerEvent, RenderEvent, SceneEvent } from '../constant';
import { LayerRender } from './layer-render';
export var LineRender = /*#__PURE__*/function (_LayerRender) {
  _inherits(LineRender, _LayerRender);
  var _super = _createSuper(LineRender);
  function LineRender() {
    var _this;
    _classCallCheck(this, LineRender);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    _this.onMouseMove = function (e) {
      _this.emit(RenderEvent.Mousemove, e);
    };
    _this.onMouseOut = function (e) {
      _this.emit(RenderEvent.Mouseout, e);
    };
    _this.onMouseDown = function (e) {
      _this.emit(RenderEvent.Dragstart, e);
    };
    _this.onDragging = function (e) {
      _this.emit(RenderEvent.Dragging, e);
    };
    _this.onDragEnd = debounce(function (e) {
      _this.emit(RenderEvent.Dragend, e);
    }, 0);
    _this.onUnClick = function (e) {
      _this.emit(RenderEvent.UnClick, e);
    };
    _this.enableHover = function () {
      var _this$layers$, _this$layers$2;
      _this.disableHover();
      (_this$layers$ = _this.layers[0]) === null || _this$layers$ === void 0 ? void 0 : _this$layers$.on(LayerEvent.Mousemove, _this.onMouseMove);
      (_this$layers$2 = _this.layers[0]) === null || _this$layers$2 === void 0 ? void 0 : _this$layers$2.on(LayerEvent.Mouseout, _this.onMouseOut);
    };
    _this.disableHover = function () {
      var _this$layers$3, _this$layers$4;
      (_this$layers$3 = _this.layers[0]) === null || _this$layers$3 === void 0 ? void 0 : _this$layers$3.off(LayerEvent.Mousemove, _this.onMouseMove);
      (_this$layers$4 = _this.layers[0]) === null || _this$layers$4 === void 0 ? void 0 : _this$layers$4.off(LayerEvent.Mouseout, _this.onMouseOut);
    };
    return _this;
  }
  _createClass(LineRender, [{
    key: "initLayers",
    value: function initLayers() {
      var _this$style = this.style,
        normal = _this$style.normal,
        hover = _this$style.hover,
        active = _this$style.active,
        style = _this$style.style,
        options = _this$style.options;
      var layer = new LineLayer(options !== null && options !== void 0 ? options : {}).source(featureCollection([])).size('isHover*isActive', function (isHover, isActive) {
        return isActive ? active.size : isHover ? hover.size : normal.size;
      }).color('isHover*isActive', function (isHover, isActive) {
        return isActive ? active.color : isHover ? hover.color : normal.color;
      }).shape('line').style(style);
      return [layer];
    }
  }, {
    key: "enableDrag",
    value: function enableDrag() {
      this.disableDrag();
      this.layers[0].on(LayerEvent.Mousedown, this.onMouseDown);
      this.scene.on(SceneEvent.Dragging, this.onDragging);
      this.scene.on(SceneEvent.Mouseup, this.onDragEnd);
      this.scene.on(SceneEvent.Dragend, this.onDragEnd);
    }
  }, {
    key: "disableDrag",
    value: function disableDrag() {
      this.layers[0].off(LayerEvent.Mousedown, this.onMouseDown);
      this.scene.off(SceneEvent.Dragging, this.onDragging);
      this.scene.off(SceneEvent.Mouseup, this.onDragEnd);
      this.scene.off(SceneEvent.Dragend, this.onDragEnd);
    }
  }, {
    key: "enableUnClick",
    value: function enableUnClick() {
      this.disableUnClick();
      this.layers[0].on(LayerEvent.UnClick, this.onUnClick);
    }
  }, {
    key: "disableUnClick",
    value: function disableUnClick() {
      this.layers[0].off(LayerEvent.UnClick, this.onUnClick);
    }
  }]);
  return LineRender;
}(LayerRender);
