// @antv/l7-draw/lib/utils/style.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSingleColorStyle = getSingleColorStyle;
/**
 * 获取单个颜色的 style 供 draw 实例使用
 * @param color
 * @returns
 */
function getSingleColorStyle(color) {
  return {
    point: {
      normal: {
        color: color
      },
      hover: {
        color: color
      },
      active: {
        color: color
      }
    },
    line: {
      normal: {
        color: color
      },
      hover: {
        color: color
      },
      active: {
        color: color
      }
    },
    polygon: {
      normal: {
        color: color
      },
      hover: {
        color: color
      },
      active: {
        color: color
      },
      style: {
        // @ts-ignore
        opacity: 0.2
      }
    },
    text: {
      normal: {
        color: color
      },
      active: {
        color: color
      }
    },
    midPoint: {
      normal: {
        color: color
      }
    },
    dashLine: {
      normal: {
        color: color
      }
    }
  };
}
