// @antv/l7-draw/lib/control/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DrawTypeList = exports.DrawTypeAttrName = exports.DrawInstanceMap = exports.DrawIconMap = exports.DrawControlClassName = exports.DrawBtnClassName = exports.DrawBtnActiveClassName = void 0;
var _drawer = require("../drawer");
var DrawTypeList = exports.DrawTypeList = ['point', 'line', 'polygon', 'rect', 'circle', 'clear'];
var DrawIconMap = exports.DrawIconMap = {
  point: '#l7draw-point',
  line: '#l7draw-line',
  polygon: '#l7draw-polygon',
  rect: '#l7draw-rect',
  circle: '#l7draw-circle',
  clear: '#l7draw-qingkong'
};
var DrawInstanceMap = exports.DrawInstanceMap = {
  point: _drawer.DrawPoint,
  line: _drawer.DrawLine,
  polygon: _drawer.DrawPolygon,
  rect: _drawer.DrawRect,
  circle: _drawer.DrawCircle
};
var DrawTypeAttrName = exports.DrawTypeAttrName = 'data-draw-type';
var DrawControlClassName = exports.DrawControlClassName = 'l7-draw-control';
var DrawBtnClassName = exports.DrawBtnClassName = "".concat(DrawControlClassName, "__btn");
var DrawBtnActiveClassName = exports.DrawBtnActiveClassName = "".concat(DrawBtnClassName, "--active");
