// @antv/l7-draw/es/mode/polygon-mode.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _get() { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get.bind(); } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(arguments.length < 3 ? target : receiver); } return desc.value; }; } return _get.apply(this, arguments); }
function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }
import { booleanClockwise, coordAll, featureCollection, lineString } from '@turf/turf';
import { cloneDeep, first, isEqual, last } from 'lodash';
import { DEFAULT_AREA_OPTIONS, DrawEvent, RenderEvent } from '../constant';
import { calcAreaText, createPointFeature, createPolygonFeature, getPosition, isSameFeature, updateTargetFeature } from '../utils';
import { LineMode } from './line-mode';
import { DEFAULT_POLYGON_HELPER_CONFIG } from '../constant/helper';
export var PolygonMode = /*#__PURE__*/function (_LineMode) {
  _inherits(PolygonMode, _LineMode);
  var _super = _createSuper(PolygonMode);
  function PolygonMode() {
    _classCallCheck(this, PolygonMode);
    return _super.apply(this, arguments);
  }
  _createClass(PolygonMode, [{
    key: "dragItem",
    get: function get() {
      return this.dragPolygon;
    }
  }, {
    key: "editItem",
    get: function get() {
      return this.editPolygon;
    }
    /**
     * 获取polygon类型对应的render
     * @protected
     */
  }, {
    key: "polygonRender",
    get: function get() {
      return this.render.polygon;
    }
    /**
     * 获取正在被拖拽的线
     * @protected
     */
  }, {
    key: "dragPolygon",
    get: function get() {
      return this.getPolygonData().find(function (feature) {
        return feature.properties.isDrag;
      });
    }
    /**
     * 正在绘制的线
     * @protected
     */
  }, {
    key: "drawPolygon",
    get: function get() {
      return this.getPolygonData().find(function (feature) {
        return feature.properties.isDraw;
      });
    }
    /**
     * 当前高亮的结点
     * @protected
     */
  }, {
    key: "editPolygon",
    get: function get() {
      return this.getPolygonData().find(function (feature) {
        var _feature$properties = feature.properties,
          isActive = _feature$properties.isActive,
          isDraw = _feature$properties.isDraw;
        return !isDraw && isActive;
      });
    }
    /**
     * 当前悬停的线
     * @protected
     */
  }, {
    key: "hoverPolygon",
    get: function get() {
      return this.getPolygonData().find(function (feature) {
        return feature.properties.isHover;
      });
    }
  }, {
    key: "getDragPolygon",
    value: function getDragPolygon() {
      return this.dragPolygon;
    }
  }, {
    key: "getDrawPolygon",
    value: function getDrawPolygon() {
      return this.drawPolygon;
    }
  }, {
    key: "getEditPolygon",
    value: function getEditPolygon() {
      return this.editPolygon;
    }
  }, {
    key: "getRenderTypes",
    value: function getRenderTypes() {
      return ['polygon', 'line', 'dashLine', 'midPoint', 'point', 'text'];
    }
  }, {
    key: "getData",
    value: function getData() {
      return this.getPolygonData();
    }
  }, {
    key: "getMainLayer",
    value: function getMainLayer() {
      var _this$render$polygon$, _this$render$polygon;
      return (_this$render$polygon$ = (_this$render$polygon = this.render.polygon) === null || _this$render$polygon === void 0 ? void 0 : _this$render$polygon.getLayers()) !== null && _this$render$polygon$ !== void 0 ? _this$render$polygon$ : [];
    }
  }, {
    key: "getDefaultOptions",
    value: function getDefaultOptions(options) {
      var newOptions = _objectSpread(_objectSpread({}, _get(_getPrototypeOf(PolygonMode.prototype), "getDefaultOptions", this).call(this, options)), {}, {
        areaOptions: false,
        helper: cloneDeep(DEFAULT_POLYGON_HELPER_CONFIG)
      });
      if (options.areaOptions) {
        newOptions.areaOptions = _objectSpread(_objectSpread({}, DEFAULT_AREA_OPTIONS), options.areaOptions);
      }
      return newOptions;
    }
  }, {
    key: "getAreaTexts",
    value: function getAreaTexts(polygons) {
      var areaOptions = this.options.areaOptions;
      if (!areaOptions) {
        return [];
      }
      var format = areaOptions.format,
        showWhen = areaOptions.showWhen;
      var textList = [];
      var polygonData = polygons.filter(function (feature) {
        return feature.geometry.coordinates[0].length >= 4;
      });
      if (showWhen.includes('active')) {
        polygonData.filter(function (feature) {
          return feature.properties.isActive;
        }).forEach(function (feature) {
          textList.push(calcAreaText(feature, {
            format: format
          }, {
            isActive: true
          }));
        });
      }
      if (showWhen.includes('normal')) {
        polygonData.filter(function (feature) {
          return !feature.properties.isActive;
        }).forEach(function (feature) {
          textList.push(calcAreaText(feature, {
            format: format
          }, {
            isActive: false
          }));
        });
      }
      return textList;
    }
  }, {
    key: "getAllTexts",
    value: function getAllTexts() {
      return [].concat(_toConsumableArray(_get(_getPrototypeOf(PolygonMode.prototype), "getAllTexts", this).call(this)), _toConsumableArray(this.getAreaTexts(this.getPolygonData())));
    }
    /**
     * 获取线数据
     */
  }, {
    key: "getPolygonData",
    value: function getPolygonData() {
      return this.source.getRenderData('polygon');
    }
    /**
     * 设置线数据
     * @param data
     */
  }, {
    key: "setPolygonData",
    value: function setPolygonData(data) {
      return this.source.setRenderData('polygon', data);
    }
  }, {
    key: "handleCreatePolygon",
    value: function handleCreatePolygon(points, line) {
      var newPolygon = createPolygonFeature(points, {
        nodes: points,
        line: line,
        isActive: true,
        isDraw: true
      });
      this.setPolygonData(function (features) {
        return updateTargetFeature({
          target: newPolygon,
          data: [].concat(_toConsumableArray(features), [newPolygon]),
          otherHandler: function otherHandler(feature) {
            feature.properties.isActive = false;
          }
        });
      });
      return newPolygon;
    }
  }, {
    key: "handlePolygonUnClick",
    value: function handlePolygonUnClick(polygon) {
      this.clearActivePolygon();
      return polygon;
    }
  }, {
    key: "handlePolygonHover",
    value: function handlePolygonHover(polygon) {
      this.setCursor('polygonHover');
      if (!isSameFeature(polygon, this.hoverPolygon)) {
        this.setPolygonData(function (features) {
          return updateTargetFeature({
            target: polygon,
            data: features,
            targetHandler: function targetHandler(feature) {
              feature.properties.isHover = true;
            },
            otherHandler: function otherHandler(feature) {
              feature.properties.isHover = false;
            }
          });
        });
      }
      return polygon;
    }
  }, {
    key: "handlePolygonUnHover",
    value: function handlePolygonUnHover() {
      this.resetCursor();
      this.setPolygonData(function (features) {
        return features.map(function (feature) {
          feature.properties.isHover = false;
          return feature;
        });
      });
    }
  }, {
    key: "handlePolygonDragStart",
    value: function handlePolygonDragStart(polygon) {
      this.setActivePolygon(polygon, {
        isDrag: true
      });
      this.scene.setMapStatus({
        dragEnable: false
      });
      this.setCursor('polygonDrag');
      return polygon;
    }
  }, {
    key: "bindPolygonRenderEvent",
    value: function bindPolygonRenderEvent() {
      var _this$polygonRender, _this$polygonRender2, _this$polygonRender3, _this$polygonRender4;
      (_this$polygonRender = this.polygonRender) === null || _this$polygonRender === void 0 ? void 0 : _this$polygonRender.on(RenderEvent.UnClick, this.onPolygonUnClick.bind(this));
      (_this$polygonRender2 = this.polygonRender) === null || _this$polygonRender2 === void 0 ? void 0 : _this$polygonRender2.on(RenderEvent.Mousemove, this.onPolygonHover.bind(this));
      (_this$polygonRender3 = this.polygonRender) === null || _this$polygonRender3 === void 0 ? void 0 : _this$polygonRender3.on(RenderEvent.Mouseout, this.onPolygonUnHover.bind(this));
      (_this$polygonRender4 = this.polygonRender) === null || _this$polygonRender4 === void 0 ? void 0 : _this$polygonRender4.on(RenderEvent.Dragstart, this.onPolygonDragStart.bind(this));
      // this.polygonRender?.on(
      //   RenderEvent.dragging,
      //   this.onPolygonDragging.bind(this),
      // );
      // this.polygonRender?.on(
      //   RenderEvent.dragend,
      //   this.onPolygonDragEnd.bind(this),
      // );
    }
  }, {
    key: "syncPolygonNodes",
    value: function syncPolygonNodes(polygon, nodes) {
      var positions = coordAll(featureCollection([].concat(_toConsumableArray(nodes), [first(nodes)])));
      var _polygon$properties = polygon.properties,
        isDraw = _polygon$properties.isDraw,
        line = _polygon$properties.line;
      polygon.properties.nodes = nodes;
      polygon.geometry.coordinates = [booleanClockwise(lineString(positions)) ? positions : positions.reverse()];
      this.setPolygonData(function (features) {
        return features.map(function (feature) {
          if (isSameFeature(feature, polygon)) {
            return polygon;
          }
          return feature;
        });
      });
      if (!isSameFeature(this.drawPolygon, polygon)) {
        var oldLineNodes = line.properties.nodes;
        var newLineNodes = _toConsumableArray(nodes);
        var firstNode = first(newLineNodes);
        if (oldLineNodes.length === nodes.length) {
          newLineNodes.push(createPointFeature(firstNode.geometry.coordinates));
        } else {
          newLineNodes.push(last(oldLineNodes));
        }
        var lastNode = last(newLineNodes);
        if (!isEqual(firstNode.geometry.coordinates, lastNode.geometry.coordinates)) {
          lastNode.geometry.coordinates = cloneDeep(firstNode.geometry.coordinates);
        }
        this.syncLineNodes(line, newLineNodes);
      }
      return polygon;
    }
  }, {
    key: "setActivePolygon",
    value: function setActivePolygon(polygon) {
      var properties = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      this.setActiveLine(polygon.properties.line, properties);
      this.setPolygonData(function (features) {
        return updateTargetFeature({
          target: polygon,
          data: features,
          targetHandler: function targetHandler(feature) {
            feature.properties = _objectSpread(_objectSpread({}, feature.properties), {}, {
              isDraw: false,
              isActive: true,
              isDrag: false,
              isHover: false
            }, properties);
          },
          otherHandler: function otherHandler(feature) {
            feature.properties = _objectSpread(_objectSpread({}, feature.properties), {}, {
              isDraw: false,
              isActive: false,
              isDrag: false
            });
          }
        });
      });
      this.setTextData(this.getAllTexts());
      return polygon;
    }
  }, {
    key: "clearActivePolygon",
    value: function clearActivePolygon() {
      this.setPolygonData(function (features) {
        return features.map(function (feature) {
          feature.properties = _objectSpread(_objectSpread({}, feature.properties), {}, {
            isDrag: false,
            isActive: false,
            isHover: false
          });
          return feature;
        });
      });
      this.clearActiveLine();
    }
  }, {
    key: "onLineDragStart",
    value: function onLineDragStart(e) {
      var line = _get(_getPrototypeOf(PolygonMode.prototype), "onLineDragStart", this).call(this, e);
      var polygon = this.getPolygonData().find(function (feature) {
        return isSameFeature(feature.properties.line, line);
      });
      if (polygon) {
        this.setActivePolygon(polygon, {
          isDrag: true
        });
        this.emit(DrawEvent.DragStart, polygon, this.getPolygonData());
      }
      return line;
    }
  }, {
    key: "onLineUnClick",
    value: function onLineUnClick(e) {
      return this.editLine;
    }
  }, {
    key: "onLineDragEnd",
    value: function onLineDragEnd(e) {
      var feature = _get(_getPrototypeOf(PolygonMode.prototype), "onLineDragEnd", this).call(this, e);
      var dragPolygon = this.dragPolygon;
      if (feature && dragPolygon) {
        dragPolygon.properties.isDrag = false;
        this.emit(DrawEvent.DragEnd, dragPolygon, this.getPolygonData());
        this.emit(DrawEvent.Edit, dragPolygon, this.getPolygonData());
      }
      return feature;
    }
  }, {
    key: "onPolygonUnClick",
    value: function onPolygonUnClick(e) {
      var editPolygon = this.editPolygon;
      if (!editPolygon) {
        return;
      }
      return this.handlePolygonUnClick(editPolygon);
    }
  }, {
    key: "onPolygonHover",
    value: function onPolygonHover(e) {
      if (this.drawPolygon) {
        return;
      }
      if (!this.dragPolygon) {
        this.setHelper('polygonHover');
      }
      return this.handlePolygonHover(e.feature);
    }
  }, {
    key: "onPolygonUnHover",
    value: function onPolygonUnHover(e) {
      if (this.drawPolygon) {
        return;
      }
      this.setHelper(this.addable ? 'draw' : null);
      return this.handlePolygonUnHover();
    }
  }, {
    key: "onPolygonDragStart",
    value: function onPolygonDragStart(e) {
      if (!this.options.editable || this.drawPolygon) {
        return;
      }
      var polygon = e.feature;
      this.previousPosition = getPosition(e);
      this.setHelper('polygonDrag');
      this.emit(DrawEvent.DragStart, polygon, this.getPolygonData());
      return this.handlePolygonDragStart(polygon);
    }
  }, {
    key: "onMidPointClick",
    value: function onMidPointClick(e) {
      var editPolygon = this.editPolygon;
      var feature = _get(_getPrototypeOf(PolygonMode.prototype), "onMidPointClick", this).call(this, e);
      if (feature && editPolygon) {
        var lineNodes = editPolygon.properties.line.properties.nodes;
        this.syncPolygonNodes(editPolygon, lineNodes.slice(0, lineNodes.length - 1));
        this.setActivePolygon(editPolygon);
      }
      return feature;
    }
  }, {
    key: "onPointDragEnd",
    value: function onPointDragEnd(e) {
      var editPolygon = this.editPolygon;
      var feature = _get(_getPrototypeOf(PolygonMode.prototype), "onPointDragEnd", this).call(this, e);
      if (feature && editPolygon) {
        this.emit(DrawEvent.Edit, editPolygon, this.getPolygonData());
      }
      return feature;
    }
  }, {
    key: "enablePolygonRenderAction",
    value: function enablePolygonRenderAction() {
      var _this$polygonRender5;
      var editable = this.options.editable;
      (_this$polygonRender5 = this.polygonRender) === null || _this$polygonRender5 === void 0 ? void 0 : _this$polygonRender5.enableUnClick();
      if (editable) {
        var _this$polygonRender6, _this$polygonRender7;
        (_this$polygonRender6 = this.polygonRender) === null || _this$polygonRender6 === void 0 ? void 0 : _this$polygonRender6.enableHover();
        (_this$polygonRender7 = this.polygonRender) === null || _this$polygonRender7 === void 0 ? void 0 : _this$polygonRender7.enableDrag();
      }
    }
  }, {
    key: "disablePolygonRenderAction",
    value: function disablePolygonRenderAction() {
      var _this$polygonRender8, _this$polygonRender9, _this$polygonRender10;
      (_this$polygonRender8 = this.polygonRender) === null || _this$polygonRender8 === void 0 ? void 0 : _this$polygonRender8.disableUnClick();
      (_this$polygonRender9 = this.polygonRender) === null || _this$polygonRender9 === void 0 ? void 0 : _this$polygonRender9.disableHover();
      (_this$polygonRender10 = this.polygonRender) === null || _this$polygonRender10 === void 0 ? void 0 : _this$polygonRender10.disableDrag();
    }
  }, {
    key: "bindEnableEvent",
    value: function bindEnableEvent() {
      _get(_getPrototypeOf(PolygonMode.prototype), "bindEnableEvent", this).call(this);
      this.enableSceneRenderAction();
      this.enablePointRenderAction();
      this.enableLineRenderAction();
      this.enableMidPointRenderAction();
      this.enablePolygonRenderAction();
    }
  }, {
    key: "unbindEnableEvent",
    value: function unbindEnableEvent() {
      _get(_getPrototypeOf(PolygonMode.prototype), "unbindEnableEvent", this).call(this);
      this.disableSceneRenderAction();
      this.disablePointRenderAction();
      this.disableLineRenderAction();
      this.disableMidPointRenderAction();
      this.disablePolygonRenderAction();
    }
  }, {
    key: "bindThis",
    value: function bindThis() {
      _get(_getPrototypeOf(PolygonMode.prototype), "bindThis", this).call(this);
      this.bindPointRenderEvent = this.bindPointRenderEvent.bind(this);
      this.bindSceneEvent = this.bindSceneEvent.bind(this);
      this.bindLineRenderEvent = this.bindLineRenderEvent.bind(this);
      this.bindMidPointRenderEvent = this.bindMidPointRenderEvent.bind(this);
      this.bindPolygonRenderEvent = this.bindPolygonRenderEvent.bind(this);
    }
  }, {
    key: "setActiveFeature",
    value: function setActiveFeature(target) {
      var targetFeature = this.getTargetFeature(target);
      if (targetFeature) {
        this.setActivePolygon(targetFeature);
      } else {
        this.clearActivePolygon();
      }
    }
  }, {
    key: "resetFeatures",
    value: function resetFeatures() {
      var features = this.getPolygonData();
      if (this.drawPolygon) {
        features = features.filter(function (feature) {
          return !feature.properties.isDraw;
        });
        this.source.setData({
          point: [],
          dashLine: [],
          midPoint: []
        });
        this.setLineData(function (features) {
          return features.filter(function (feature) {
            return !feature.properties.isDraw;
          });
        });
        this.setTextData(function (features) {
          return features.filter(function (feature) {
            return !feature.properties.isActive;
          });
        });
      }
      if (this.editPolygon) {
        this.handlePolygonUnClick(this.editPolygon);
      }
      this.setPolygonData(features.map(function (feature) {
        feature.properties = _objectSpread(_objectSpread({}, feature.properties), {}, {
          isDrag: false,
          isActive: false,
          isHover: false
        });
        return feature;
      }));
    }
  }]);
  return PolygonMode;
}(LineMode);
