// @antv/l7-layers/es/geometry/models/billboard.js
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _regeneratorRuntime from "@babel/runtime/regenerator";
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
import { AttributeType, gl } from '@antv/l7-core';
import BaseModel from "../../core/BaseModel";
/* babel-plugin-inline-import '../shaders/billboard_frag.glsl' */
var planeFrag = "layout(std140) uniform commonUniforms {\n vec2 u_size;\n float u_raisingHeight;\n float u_rotation;\n float u_opacity;\n};\n\nuniform sampler2D u_texture;\n\nin vec2 v_uv;\nout vec4 outputColor;\n\n#pragma include \"picking\"\nvoid main() {\n  outputColor = texture(SAMPLER_2D(u_texture), vec2(v_uv.x, 1.0 - v_uv.y));\n  outputColor.a *= u_opacity;\n  outputColor = filterColor(outputColor);\n}\n";
/* babel-plugin-inline-import '../shaders/billboard_vert.glsl' */
var planeVert = "layout(location = 0) in vec3 a_Position;\nlayout(location = 11) in vec3 a_Extrude;\nlayout(location = 14) in vec2 a_Uv;\n\nlayout(std140) uniform commonUniforms {\n vec2 u_size;\n float u_raisingHeight;\n float u_rotation;\n float u_opacity;\n};\n\nout vec2 v_uv;\n\n#pragma include \"projection\"\n#pragma include \"picking\"\n#pragma include \"rotation_2d\"\nvoid main() {\n   vec3 extrude = a_Extrude;\n   v_uv = a_Uv;\n   float raiseHeight = u_raisingHeight;\n   if(u_CoordinateSystem == COORDINATE_SYSTEM_LNGLAT || u_CoordinateSystem == COORDINATE_SYSTEM_LNGLAT_OFFSET) {\n      float mapboxZoomScale = 4.0/pow(2.0, 21.0 - u_Zoom);\n      raiseHeight = u_raisingHeight * mapboxZoomScale;\n   }\n\n   // \u8BA1\u7B97\u7ECF\u7EAC\u5EA6\u70B9\u4F4D\u5750\u6807\n   vec4 project_pos = project_position(vec4(a_Position.xy, 0.0, 1.0));\n\n   // \u8BA1\u7B97\u7ED5 z \u8F74\u65CB\u8F6C\u540E\u7684\u504F\u79FB\n   vec2 offsetXY = project_pixel(rotate_matrix(vec2(extrude.x * u_size.x, 0.0),u_rotation));\n   // \u7ED5 z \u8F74\u65CB\u8F6C\n   float x = project_pos.x + offsetXY.x;\n   float y = project_pos.y + offsetXY.y;\n   // z \u8F74\u4E0D\u53C2\u4E0E\u65CB\u8F6C\n   float z = project_pixel(extrude.y * u_size.y + raiseHeight);\n\n   gl_Position = project_common_position_to_clipspace_v2(vec4(x , y, z , 1.0));\n\n   setPickingColor(a_PickingColor);\n}\n";
import { ShaderLocation } from "../../core/CommonStyleAttribute";
var BillBoardModel = /*#__PURE__*/function (_BaseModel) {
  _inherits(BillBoardModel, _BaseModel);
  var _super = _createSuper(BillBoardModel);
  function BillBoardModel() {
    var _this;
    _classCallCheck(this, BillBoardModel);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    _defineProperty(_assertThisInitialized(_this), "radian", 0);
    // 旋转的弧度
    _defineProperty(_assertThisInitialized(_this), "planeGeometryTriangulation", function () {
      var _ref = _this.layer.getLayerConfig(),
        _ref$center = _ref.center,
        center = _ref$center === void 0 ? [120, 30] : _ref$center;
      return {
        size: 4,
        indices: [0, 1, 2, 2, 3, 0],
        vertices: [].concat(_toConsumableArray(center), [1, 1], _toConsumableArray(center), [0, 1], _toConsumableArray(center), [0, 0], _toConsumableArray(center), [1, 0])
      };
    });
    return _this;
  }
  _createClass(BillBoardModel, [{
    key: "getUninforms",
    value: function getUninforms() {
      var commoninfo = this.getCommonUniformsInfo();
      var attributeInfo = this.getUniformsBufferInfo(this.getStyleAttribute());
      this.updateStyleUnifoms();
      return _objectSpread(_objectSpread({}, commoninfo.uniformsOption), attributeInfo.uniformsOption);
    }
  }, {
    key: "getCommonUniformsInfo",
    value: function getCommonUniformsInfo() {
      var _ref2 = this.layer.getLayerConfig(),
        opacity = _ref2.opacity,
        _ref2$width = _ref2.width,
        width = _ref2$width === void 0 ? 1 : _ref2$width,
        _ref2$height = _ref2.height,
        height = _ref2$height === void 0 ? 1 : _ref2$height,
        _ref2$raisingHeight = _ref2.raisingHeight,
        raisingHeight = _ref2$raisingHeight === void 0 ? 0 : _ref2$raisingHeight;

      /**
       *               rotateFlag
       * DEFAULT            1
       * MAPBOX           1
       * GAODE2.x         -1
       * GAODE1.x         -1
       */
      var rotateFlag = 1;
      if (this.mapService.version === 'GAODE2.x' || this.mapService.version === 'GAODE1.x') {
        rotateFlag = -1;
      }
      // 控制图标的旋转角度（绕 Z 轴旋转）
      this.radian = rotateFlag * Math.PI * (this.mapService.getRotation() % 360) / 180;
      var commonOptions = {
        u_size: [width, height],
        u_raisingHeight: Number(raisingHeight),
        u_rotation: this.radian,
        u_opacity: opacity || 1,
        u_texture: this.texture
      };
      this.textures = [this.texture];
      var commonBufferInfo = this.getUniformsBufferInfo(commonOptions);
      return commonBufferInfo;
    }
  }, {
    key: "clearModels",
    value: function clearModels() {
      var _this$texture;
      (_this$texture = this.texture) === null || _this$texture === void 0 || _this$texture.destroy();
    }
  }, {
    key: "initModels",
    value: function () {
      var _initModels = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        var _ref3, drawCanvas, createTexture2D, model;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _ref3 = this.layer.getLayerConfig(), drawCanvas = _ref3.drawCanvas;
              createTexture2D = this.rendererService.createTexture2D;
              this.texture = createTexture2D({
                height: 0,
                width: 0
              });
              if (drawCanvas) {
                this.updateTexture(drawCanvas);
              }
              this.initUniformsBuffer();
              _context.next = 7;
              return this.layer.buildLayerModel({
                moduleName: 'geometryBillboard',
                vertexShader: planeVert,
                fragmentShader: planeFrag,
                triangulation: this.planeGeometryTriangulation,
                inject: this.getInject(),
                primitive: gl.TRIANGLES,
                depth: {
                  enable: true
                }
              });
            case 7:
              model = _context.sent;
              return _context.abrupt("return", [model]);
            case 9:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function initModels() {
        return _initModels.apply(this, arguments);
      }
      return initModels;
    }()
  }, {
    key: "buildModels",
    value: function () {
      var _buildModels = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              return _context2.abrupt("return", this.initModels());
            case 1:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function buildModels() {
        return _buildModels.apply(this, arguments);
      }
      return buildModels;
    }()
  }, {
    key: "updateTexture",
    value: function updateTexture(drawCanvas) {
      var createTexture2D = this.rendererService.createTexture2D;
      var _ref4 = this.layer.getLayerConfig(),
        _ref4$canvasWidth = _ref4.canvasWidth,
        canvasWidth = _ref4$canvasWidth === void 0 ? 1 : _ref4$canvasWidth,
        _ref4$canvasHeight = _ref4.canvasHeight,
        canvasHeight = _ref4$canvasHeight === void 0 ? 1 : _ref4$canvasHeight;
      var canvas = document.createElement('canvas');
      canvas.width = canvasWidth;
      canvas.height = canvasHeight;
      var ctx = canvas.getContext('2d');
      if (ctx) {
        drawCanvas(canvas);
        this.texture = createTexture2D({
          data: canvas,
          width: canvas.width,
          height: canvas.height,
          wrapS: gl.CLAMP_TO_EDGE,
          wrapT: gl.CLAMP_TO_EDGE
        });
        this.layerService.reRender();
      }
    }
  }, {
    key: "registerBuiltinAttributes",
    value: function registerBuiltinAttributes() {
      this.styleAttributeService.registerStyleAttribute({
        name: 'extrude',
        type: AttributeType.Attribute,
        descriptor: {
          name: 'a_Extrude',
          shaderLocation: ShaderLocation.EXTRUDE,
          buffer: {
            usage: gl.DYNAMIC_DRAW,
            data: [],
            type: gl.FLOAT
          },
          size: 3,
          update: function update(feature, featureIdx, vertex, attributeIdx) {
            var extrude = [1, 1, 0, -1, 1, 0, -1, -1, 0, 1, -1, 0];
            var extrudeIndex = attributeIdx % 4 * 3;
            return [extrude[extrudeIndex], extrude[extrudeIndex + 1], extrude[extrudeIndex + 2]];
          }
        }
      });
      this.styleAttributeService.registerStyleAttribute({
        name: 'uv',
        type: AttributeType.Attribute,
        descriptor: {
          name: 'a_Uv',
          shaderLocation: ShaderLocation.UV,
          buffer: {
            usage: gl.DYNAMIC_DRAW,
            data: [],
            type: gl.FLOAT
          },
          size: 2,
          update: function update(feature, featureIdx, vertex) {
            return [vertex[2], vertex[3]];
          }
        }
      });
    }
  }]);
  return BillBoardModel;
}(BaseModel);
export { BillBoardModel as default };
