// @antv/l7-draw/es/constant/drawer.js
import { followCursor } from 'tippy.js';
export var DEFAULT_DISTANCE_OPTIONS = {
  showTotalDistance: false,
  showDashDistance: true,
  showWhen: ['normal', 'active'],
  format: function format(meters) {
    if (meters >= 1000) {
      return +(meters / 1000).toFixed(2) + 'km';
    } else {
      return +meters.toFixed(2) + 'm';
    }
  }
};
export var DEFAULT_AREA_OPTIONS = {
  showWhen: ['normal', 'active'],
  format: function format(squareMeters) {
    return squareMeters > 1000000 ? "".concat(+(squareMeters / 1000000).toFixed(2), "km\xB2") : "".concat(+squareMeters.toFixed(2), "m\xB2");
  }
};
export var DEFAULT_HISTORY_CONFIG = {
  maxSize: 100
};
export var DEFAULT_KEYBOARD_CONFIG = {
  remove: ['del', 'backspace'],
  revert: ['command+z', 'ctrl+z'],
  redo: ['command+shift+z', 'ctrl+shift+z']
};
export var DEFAULT_POPUP_CONFIG = {
  allowHTML: true,
  followCursor: true,
  placement: 'bottom-start',
  offset: [-10, 20],
  plugins: [followCursor],
  trigger: 'manual',
  hideOnClick: false,
  theme: 'light'
};
export var DEFAULT_ADSORB_CONFIG = {
  data: 'drawData',
  pointAdsorbPixel: 12,
  lineAdsorbPixel: 10
};
