// @antv/l7-draw/es/constant/style.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
export var NORMAL_COLOR = '#1990FF';
export var ACTIVE_COLOR = '#ED9D48';
export var DEFAULT_COMMON_OPTIONS = {
  blend: 'normal',
  pickingBuffer: 5
};
export var DEFAULT_POINT_NORMAL_STYLE = {
  color: NORMAL_COLOR,
  shape: 'circle',
  size: 6
};
export var DEFAULT_NODE_NORMAL_STYLE = {
  color: ACTIVE_COLOR,
  shape: 'circle',
  size: 6
};
export var DEFAULT_MID_POINT_STYLE = {
  shape: 'circle',
  size: 6,
  color: ACTIVE_COLOR
};
export var DEFAULT_POINT_STYLE = {
  options: DEFAULT_COMMON_OPTIONS,
  normal: DEFAULT_POINT_NORMAL_STYLE,
  hover: _objectSpread(_objectSpread({}, DEFAULT_POINT_NORMAL_STYLE), {}, {
    size: 8
  }),
  active: _objectSpread(_objectSpread({}, DEFAULT_POINT_NORMAL_STYLE), {}, {
    size: 8,
    color: ACTIVE_COLOR
  }),
  style: {
    stroke: '#ffffff',
    strokeWidth: 2
  }
};
export var DEFAULT_NODE_STYLE = {
  options: DEFAULT_COMMON_OPTIONS,
  normal: DEFAULT_NODE_NORMAL_STYLE,
  hover: _objectSpread(_objectSpread({}, DEFAULT_NODE_NORMAL_STYLE), {}, {
    size: 8
  }),
  active: _objectSpread(_objectSpread({}, DEFAULT_NODE_NORMAL_STYLE), {}, {
    size: 8
  }),
  style: {
    stroke: '#ffffff',
    strokeWidth: 2
  }
};
export var DEFAULT_TEXT_NORMAL_STYLE = {
  color: NORMAL_COLOR,
  size: 12
};
export var DEFAULT_STYLE = {
  point: DEFAULT_NODE_STYLE,
  line: {
    options: DEFAULT_COMMON_OPTIONS,
    normal: {
      color: NORMAL_COLOR,
      size: 2
    },
    hover: {
      color: NORMAL_COLOR,
      size: 2
    },
    active: {
      color: ACTIVE_COLOR,
      size: 2
    },
    style: {}
  },
  polygon: {
    options: DEFAULT_COMMON_OPTIONS,
    normal: {
      color: NORMAL_COLOR
    },
    hover: {
      color: NORMAL_COLOR
    },
    active: {
      color: ACTIVE_COLOR
    },
    style: {
      opacity: 0.15
    }
  },
  midPoint: {
    options: DEFAULT_COMMON_OPTIONS,
    normal: DEFAULT_MID_POINT_STYLE
  },
  dashLine: {
    options: DEFAULT_COMMON_OPTIONS,
    normal: {
      color: ACTIVE_COLOR,
      size: 2
    },
    style: {
      lineType: 'dash',
      dashArray: [6, 6]
    }
  },
  text: {
    options: DEFAULT_COMMON_OPTIONS,
    normal: DEFAULT_TEXT_NORMAL_STYLE,
    active: _objectSpread(_objectSpread({}, DEFAULT_TEXT_NORMAL_STYLE), {}, {
      color: ACTIVE_COLOR
    }),
    style: {
      fontWeight: '800',
      textOffset: [0, DEFAULT_TEXT_NORMAL_STYLE.size + 6],
      textAllowOverlap: true,
      stroke: '#ffffff',
      strokeWidth: 2
    }
  }
};
