// @antv/l7-source/es/parser/raster/ndi.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["extent", "width", "height"];
/**
 * @description: 栅格数据解析
 */

export default function rasterRgb(data, cfg) {
  var extent = cfg.extent,
    width = cfg.width,
    height = cfg.height,
    options = _objectWithoutProperties(cfg, _excluded);
  if (data.length < 2) {
    console.warn('RGB解析需要2个波段的数据');
  }
  var _ref = options.bands || [0, 1],
    _ref2 = _slicedToArray(_ref, 2),
    n = _ref2[0],
    d = _ref2[1];
  var bandsData = [data[n], data[d]];
  var ndidata = [];
  for (var i = 0; i < bandsData[0].length; i++) {
    ndidata.push((bandsData[1][i] - bandsData[0][i]) / (bandsData[1][i] + bandsData[0][i]));
  }
  var resultData = {
    _id: 1,
    dataArray: [_objectSpread(_objectSpread({
      _id: 1,
      data: ndidata,
      width: width,
      height: height
    }, options), {}, {
      coordinates: [[extent[0], extent[1]], [extent[2], extent[3]]]
    })]
  };
  return resultData;
}
