// @antv/l7-draw/lib/constant/drawer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_POPUP_CONFIG = exports.DEFAULT_KEYBOARD_CONFIG = exports.DEFAULT_HISTORY_CONFIG = exports.DEFAULT_DISTANCE_OPTIONS = exports.DEFAULT_AREA_OPTIONS = exports.DEFAULT_ADSORB_CONFIG = void 0;
var _tippy = require("tippy.js");
var DEFAULT_DISTANCE_OPTIONS = exports.DEFAULT_DISTANCE_OPTIONS = {
  showTotalDistance: false,
  showDashDistance: true,
  showWhen: ['normal', 'active'],
  format: function format(meters) {
    if (meters >= 1000) {
      return +(meters / 1000).toFixed(2) + 'km';
    } else {
      return +meters.toFixed(2) + 'm';
    }
  }
};
var DEFAULT_AREA_OPTIONS = exports.DEFAULT_AREA_OPTIONS = {
  showWhen: ['normal', 'active'],
  format: function format(squareMeters) {
    return squareMeters > 1000000 ? "".concat(+(squareMeters / 1000000).toFixed(2), "km\xB2") : "".concat(+squareMeters.toFixed(2), "m\xB2");
  }
};
var DEFAULT_HISTORY_CONFIG = exports.DEFAULT_HISTORY_CONFIG = {
  maxSize: 100
};
var DEFAULT_KEYBOARD_CONFIG = exports.DEFAULT_KEYBOARD_CONFIG = {
  remove: ['del', 'backspace'],
  revert: ['command+z', 'ctrl+z'],
  redo: ['command+shift+z', 'ctrl+shift+z']
};
var DEFAULT_POPUP_CONFIG = exports.DEFAULT_POPUP_CONFIG = {
  allowHTML: true,
  followCursor: true,
  placement: 'bottom-start',
  offset: [-10, 20],
  plugins: [_tippy.followCursor],
  trigger: 'manual',
  hideOnClick: false,
  theme: 'light'
};
var DEFAULT_ADSORB_CONFIG = exports.DEFAULT_ADSORB_CONFIG = {
  data: 'drawData',
  pointAdsorbPixel: 12,
  lineAdsorbPixel: 10
};
