// @antv/l7-draw/lib/constant/event.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SourceEvent = exports.SceneEvent = exports.RenderEvent = exports.LayerEvent = exports.DrawEvent = exports.ControlEvent = void 0;
/**
 * Drawer事件枚举
 */
var DrawEvent;
(function (DrawEvent) {
  DrawEvent["Init"] = "init";
  DrawEvent["Destroy"] = "destroy";
  DrawEvent["Enable"] = "enable";
  DrawEvent["Disable"] = "disable";
  DrawEvent["Add"] = "add";
  DrawEvent["Edit"] = "edit";
  DrawEvent["Remove"] = "remove";
  DrawEvent["Clear"] = "clear";
  DrawEvent["Change"] = "change";
  DrawEvent["DragStart"] = "dragStart";
  DrawEvent["Dragging"] = "dragging";
  DrawEvent["DragEnd"] = "dragEnd";
  DrawEvent["Select"] = "select";
  DrawEvent["AddNode"] = "addNode";
  DrawEvent["RemoveNode"] = "removeNode";
})(DrawEvent || (exports.DrawEvent = DrawEvent = {}));
/**
 * Source事件枚举
 */
var SourceEvent;
(function (SourceEvent) {
  SourceEvent["Change"] = "change";
  SourceEvent["Update"] = "update";
})(SourceEvent || (exports.SourceEvent = SourceEvent = {}));
/**
 * Render事件枚举
 */
var RenderEvent;
(function (RenderEvent) {
  RenderEvent["Click"] = "click";
  RenderEvent["UnClick"] = "unclick";
  RenderEvent["Dragstart"] = "dragstart";
  RenderEvent["Mousemove"] = "mousemove";
  RenderEvent["Mouseout"] = "mouseout";
  RenderEvent["Dragging"] = "dragging";
  RenderEvent["Dragend"] = "dragend";
  RenderEvent["DblClick"] = "dblClick";
  RenderEvent["Contextmenu"] = "contextmenu";
})(RenderEvent || (exports.RenderEvent = RenderEvent = {}));
/**
 * L7 Layer 事件名枚举
 */
var LayerEvent;
(function (LayerEvent) {
  LayerEvent["Mousedown"] = "mousedown";
  LayerEvent["Mouseup"] = "mouseup";
  LayerEvent["Click"] = "click";
  LayerEvent["UnClick"] = "unclick";
  LayerEvent["Dblclick"] = "dblclick";
  LayerEvent["Mousemove"] = "mousemove";
  LayerEvent["Mouseover"] = "mouseover";
  LayerEvent["Mouseenter"] = "mouseenter";
  LayerEvent["Mouseleave"] = "mouseleave";
  LayerEvent["Mouseout"] = "mouseout";
  LayerEvent["Contextmenu"] = "contextmenu";
})(LayerEvent || (exports.LayerEvent = LayerEvent = {}));
/**
 * L7 Scene 事件名枚举
 */
var SceneEvent;
(function (SceneEvent) {
  SceneEvent["Loaded"] = "loaded";
  SceneEvent["MapMove"] = "mapmove";
  SceneEvent["MoveStart"] = "movestart";
  SceneEvent["MoveEnd"] = "moveend";
  SceneEvent["ZoomChange"] = "zoomchange";
  SceneEvent["ZoomStart"] = "zoomstart";
  SceneEvent["ZoomEnd"] = "zoomend";
  SceneEvent["Click"] = "click";
  SceneEvent["Dblclick"] = "dblclick";
  SceneEvent["Mousemove"] = "mousemove";
  SceneEvent["Mousewheel"] = "mousewheel";
  SceneEvent["Mouseover"] = "mouseover";
  SceneEvent["Mouseout"] = "mouseout";
  SceneEvent["Mouseup"] = "mouseup";
  SceneEvent["Mousedown"] = "mousedown";
  SceneEvent["Contextmenu"] = "contextmenu";
  SceneEvent["Dragstart"] = "dragstart";
  SceneEvent["Dragging"] = "dragging";
  SceneEvent["Dragend"] = "dragend";
})(SceneEvent || (exports.SceneEvent = SceneEvent = {}));
var ControlEvent;
(function (ControlEvent) {
  ControlEvent["DrawChange"] = "drawchange";
  ControlEvent["DataChange"] = "datachange";
  ControlEvent["DrawSelect"] = "drawselect";
  ControlEvent["DrawClear"] = "drawclear";
})(ControlEvent || (exports.ControlEvent = ControlEvent = {}));
